src/Services/CheckoutService.php

```diff
@@ -1,251 +0,0 @@
-            return false;
-        }
-
-        if ($voucher = app('voucher')::validAt()->where('code', $code)->first()) {
-            return $voucher;
-        }
-
-        if ($discount = app('discount')::available()->where('code', $code)->first()) {
-            return $discount;
-        }
-
-        return false;
-    }
-
-    /**
-     * Apply discount or voucher code to cart.
-     *
-     * @param string $code
-     * @return Cart
-     */
-    public function applyCodeToCart($code, $cart)
-    {
-        $deduction = $this->findDeductionByCode($code);
-
-        if (empty($deduction)) {
-            throw new \Exception("Code {$code} is invalid.");
-        }
-
-        $cart->applyDeduction($deduction);
-
-        return $cart;
-    }
-
-    /**
-     * Apply deduction to cart.
-     *
-     * @param Model $deduction
-     * @param Cart $cart
-     * @return Cart
-     */
-    public function applyDeductionToCart($deduction, $cart)
-    {
-        // TODO - temporary ugliness, will be moved into discounts/vouchers services
-        if (is_a($deduction, app('discount'))) {
-            $cart->applyDiscount($deduction);
-        }
-
-        if (is_a($deduction, app('voucher'))) {
-            $cart->applyVoucher($deduction);
-        }
-
-        return $cart;
-    }
-
-    /**
-     * Apply discount to cart.
-     *
-     * @param Model $discount
-     * @param Cart $cart
-     * @return Cart
-     */
-    public function applyDiscountToCart($discount, $cart)
-    {
-        if (! in_array($discount->applies_to, [self::APPLICATION_ORDER, self::APPLICATION_PARTICIPANT])) {
-            return $cart;
-        }
-
-        $cart->discounts()->syncWithoutDetaching([$discount->id]);
-        $cart->total_deductions = $this->calculateCartDeductions($cart);
-        $cart->save();
-
-        return $cart;
-    }
-
-    /**
-     * Issue partial redemption voucher when total discounts are higher than total amount.
-     *
-     * @param  Cart $cart
-     * @return Model
-     */
-    public function issueCartPartialRedemptionVoucher($cart)
-    {
-        $order = $cart->order;
-
-        /** @psalm-suppress UndefinedMethod */
-        $voucher = app('voucher')::create([
-            'location_id' => $cart->location_id,
-            'customer_id' => $cart->user_id,
-            'voucher_type_id' => Cart::PARTIAL_REDEMPTION_VOUCHER_TYPE_ID,
-            'redeemable_at' => now(),
-            'amount' => $cart->total_deductions - ($cart->amount + $cart->total_taxes + $cart->total_fees),
-            'expires_at' => $cart->vouchers()->first()->expires_at,
-            'creator_id' => $cart->user_id,
-            'updater_id' => $cart->user_id,
-        ]);
-
-        $order->partial_redemption_voucher_id = $voucher->id;
-        $order->save();
-
-        return $voucher;
-    }
-
-    /**
-     * Calculate all cart deductions including items deductions.
-     *
-     * @return int
-     */
-    public function calculateCartDeductions($cart)
```

Comment:
Moved to Cart::updateTotalDeductions() as private helper, invoked after a deduction is applied.  Can be promoted later (to public method or CartInterface) if there is external need.